### ISSUE TITLE
Orchestration dies after out of memory exception 

### ISSUE THREAD

**Issue by jbarntish** [microsoft/durabletask-netherite#232]
I am not 100% sure if this is a Durable Task issue, or a Netherite issue. But the stack trace indicates Neterite is at least involved in some fashion.

We have a setup where a main orchestrator fans out over 15k records, calling sub orchestrations. The data size is relatively small, and data being passed in to the sub orchestrations is also very small. The sub orchestration does several SQL Server based call activities, and a couple of cosmos activities. Due to the speed of netherite, it actually processes so fast we have had to scale everything up. (This has been a bit of a lifeline for us, as the default provider simply does not have enough juice for our needs.)

At a certain point the orchestration just ... dies. I can get 2k sub orchestrations complete, sometimes 4k. There are no unhandled exceptions in our durable function stack that correlate to this (no sql connection exhaustion, cosmos 429s, or other unhandleds).

Not only does the orchestration die, but the func app at this point becomes completely unresponsive. Hitting DFMon endpoints via insomnia to get status never complete. The func app at this point is idle; no more executions. Stopping/restarting does not clear it. I have to destroy the func app and it's web app host and redeploy. Memory is not even close to being fully utilized (as far as I can tell). This may or may not be related, just providing all the info I can. 

The fun app trace logs go along fine:

```
Executed 'VehicleOrchestrator' (Succeeded, Id=add0af1d-409e-477e-8d0b-16733b69608e, Duration=0ms)
0e535510c29f46c4875062202ada637d:3508: Function 'VehicleOrchestrator (Orchestrator)' completed. ContinuedAsNew: False. IsReplay: False. Output: (null). State: Completed. HubName: foo. AppName: df-app. SlotName: Production. ExtensionVersion: 2.9.1. SequenceNumber: 78141. TaskEventId: -1
VehicleOrchestrator finished!
CHECKPOINT: VEHICLE COMPLETE
Executing 'VehicleOrchestrator' (Reason='(null)', Id=add0af1d-409e-477e-8d0b-16733b69608e)
Executed 'VehicleOrchestrator' (Succeeded, Id=42187ac5-a421-4fc6-99e4-d3e98d719f92, Duration=0ms)
```

And then the first error that occurs (there's maybe 20ish of these) before the orchestration stops:

> Incidentally, I found this which speaks to potential gaps in memory handling: https://github.com/Azure/durabletask/issues/629, may or may not be related.

```
EventHubsProcessor partitions/5 received internal error indication from EventProcessorHost: System.OutOfMemoryException: Exception of type 'System.OutOfMemoryException' was thrown. at 
System.GC.AllocateNewArray(IntPtr typeHandle, Int32 length, GC_ALLOC_FLAGS flags) at 
System.GC.AllocateArray[T](Int32 length, Boolean pinned) at FASTER.core.BlittableAllocator`2.AllocatePage(Int32 index) at 
FASTER.core.AllocatorBase`2.Initialize(Int64 firstValidAddress) at FASTER.core.FasterLog..ctor(FasterLogSettings logSettings, Boolean syncRecover, ILogger logger) at 
DurableTask.Netherite.Faster.FasterLog..ctor(BlobManager blobManager, NetheriteOrchestrationServiceSettings settings) in /_/src/DurableTask.Netherite/StorageLayer/Faster/FasterLog.cs:line 22 at 
DurableTask.Netherite.Faster.PartitionStorage.CreateOrRestoreAsync(Partition partition, IPartitionErrorHandler errorHandler, String inputQueueFingerprint) in /_/src/DurableTask.Netherite/StorageLayer/Faster/PartitionStorage.cs:line 103 at 
DurableTask.Netherite.Partition.CreateOrRestoreAsync(IPartitionErrorHandler errorHandler, TaskhubParameters parameters, String inputQueueFingerprint) in /_/src/DurableTask.Netherite/OrchestrationService/Partition.cs:line 124 at 
DurableTask.Netherite.EventHubsTransport.EventHubsProcessor.StartPartitionAsync(PartitionIncarnation prior) in /_/src/DurableTask.Netherite/TransportLayer/EventHubs/EventHubsProcessor.cs:line 201 at 
DurableTask.Netherite.EventHubsTransport.EventHubsProcessor.Microsoft.Azure.EventHubs.Processor.IEventProcessor.ProcessEventsAsync(PartitionContext context, IEnumerable`1 packets) in /_/src/DurableTask.Netherite/TransportLayer/EventHubs/EventHubsProcessor.cs:line 362 at 
Microsoft.Azure.EventHubs.Processor.PartitionPump.ProcessEventsAsync(IEnumerable`1 events)
```

And from then on it's a series of these: 

> Quite similar in nature to https://github.com/microsoft/durabletask-netherite/issues/231 in terms of the `Part04 !!!` messaging:

```
Part04 !!! Could not start partition in CreateOrRestoreAsync: System.ArgumentException: Destination array was not long enough. Check the destination index, length, and the array's lower bounds. (Parameter 'destinationArray')
at System.Array.Copy(Array sourceArray, Int32 sourceIndex, Array destinationArray, Int32 destinationIndex, Int32 length, Boolean reliable)
at System.Array.Copy(Array sourceArray, Array destinationArray, Int32 length)
at FASTER.core.FasterKV`2.SetRecoveryPageRanges(HybridLogCheckpointInfo recoveredHLCInfo, Int32 numPagesToPreload, Int64 fromAddress, Int64& tailAddress, Int64& headAddress, Int64& scanFromAddress)
at FASTER.core.FasterKV`2.InternalRecoverAsync(IndexCheckpointInfo recoveredICInfo, HybridLogCheckpointInfo recoveredHLCInfo, Int32 numPagesToPreload, Boolean undoNextVersion, Int64 recoverTo, CancellationToken cancellationToken)
at DurableTask.Netherite.Faster.FasterKV.RecoverAsync() in /_/src/DurableTask.Netherite/StorageLayer/Faster/FasterKV.cs:line 252
at DurableTask.Netherite.Faster.PartitionStorage.TerminationWrapper[T](Task`1 what)
at DurableTask.Netherite.Faster.PartitionStorage.CreateOrRestoreAsync(Partition partition, IPartitionErrorHandler errorHandler, String inputQueueFingerprint) in /_/src/DurableTask.Netherite/StorageLayer/Faster/PartitionStorage.cs:line 178
at DurableTask.Netherite.Partition.CreateOrRestoreAsync(IPartitionErrorHandler errorHandler, TaskhubParameters parameters, String inputQueueFingerprint) in /_/src/DurableTask.Netherite/OrchestrationService/Partition.cs:line 124 terminatePartition=True
```
```
Part04 !!! failed to start partition in EventHubsProcessor.StartPartitionAsync: System.ArgumentException: Destination array was not long enough. Check the destination index, length, and the array's lower bounds. (Parameter 'destinationArray')
at System.Array.Copy(Array sourceArray, Int32 sourceIndex, Array destinationArray, Int32 destinationIndex, Int32 length, Boolean reliable)
at System.Array.Copy(Array sourceArray, Array destinationArray, Int32 length)
at FASTER.core.FasterKV`2.SetRecoveryPageRanges(HybridLogCheckpointInfo recoveredHLCInfo, Int32 numPagesToPreload, Int64 fromAddress, Int64& tailAddress, Int64& headAddress, Int64& scanFromAddress)
at FASTER.core.FasterKV`2.InternalRecoverAsync(IndexCheckpointInfo recoveredICInfo, HybridLogCheckpointInfo recoveredHLCInfo, Int32 numPagesToPreload, Boolean undoNextVersion, Int64 recoverTo, CancellationToken cancellationToken)
at DurableTask.Netherite.Faster.FasterKV.RecoverAsync() in /_/src/DurableTask.Netherite/StorageLayer/Faster/FasterKV.cs:line 252
at DurableTask.Netherite.Faster.PartitionStorage.TerminationWrapper[T](Task`1 what)
at DurableTask.Netherite.Faster.PartitionStorage.CreateOrRestoreAsync(Partition partition, IPartitionErrorHandler errorHandler, String inputQueueFingerprint) in /_/src/DurableTask.Netherite/StorageLayer/Faster/PartitionStorage.cs:line 178
at DurableTask.Netherite.Partition.CreateOrRestoreAsync(IPartitionErrorHandler errorHandler, TaskhubParameters parameters, String inputQueueFingerprint) in /_/src/DurableTask.Netherite/OrchestrationService/Partition.cs:line 124
at DurableTask.Netherite.EventHubsTransport.EventHubsProcessor.StartPartitionAsync(PartitionIncarnation prior) in /_/src/DurableTask.Netherite/TransportLayer/EventHubs/EventHubsProcessor.cs:line 201 terminatePartition=True
```

```
Part04 !!! Faster Storage Error : loading checkpoint : System.ArgumentException: Destination array was not long enough. Check the destination index, length, and the array's lower bounds. (Parameter 'destinationArray')
at System.Array.Copy(Array sourceArray, Int32 sourceIndex, Array destinationArray, Int32 destinationIndex, Int32 length, Boolean reliable)
at System.Array.Copy(Array sourceArray, Array destinationArray, Int32 length)
at FASTER.core.FasterKV`2.SetRecoveryPageRanges(HybridLogCheckpointInfo recoveredHLCInfo, Int32 numPagesToPreload, Int64 fromAddress, Int64& tailAddress, Int64& headAddress, Int64& scanFromAddress)
at FASTER.core.FasterKV`2.InternalRecoverAsync(IndexCheckpointInfo recoveredICInfo, HybridLogCheckpointInfo recoveredHLCInfo, Int32 numPagesToPreload, Boolean undoNextVersion, Int64 recoverTo, CancellationToken cancellationToken)
at DurableTask.Netherite.Faster.FasterKV.RecoverAsync() in /_/src/DurableTask.Netherite/StorageLayer/Faster/FasterKV.cs:line 252
at DurableTask.Netherite.Faster.PartitionStorage.TerminationWrapper[T](Task`1 what)
at DurableTask.Netherite.Faster.PartitionStorage.CreateOrRestoreAsync(Partition partition, IPartitionErrorHandler errorHandler, String inputQueueFingerprint) in /_/src/DurableTask.Netherite/StorageLayer/Faster/PartitionStorage.cs:line 178
```

In terms of stack, I am running the function app on a `P2v3`, with these settings in `host.json`:

```json
{
    "version": "2.0",
    "aggregator": {
        "batchSize": 2000,
        "flushTimeout": "00:01:00"
    },
    "concurrency": {
        "dynamicConcurrencyEnabled": false,
        "snapshotPersistenceEnabled": false
    },
    "extensions": {
        "durableTask": {
            "hubName": "foo",
            "useGracefulShutdown": false,
            "maxConcurrentActivityFunctions": 30,
            "maxConcurrentOrchestratorFunctions": 10,
            "tracing": {
                "traceInputsAndOutputs": false,
                "traceReplayEvents": false
            },
            "storageProvider": {
                "type": "Netherite",
                "partitionCount": 12,
                "StorageConnectionName": "AzureWebJobsStorage",
                "EventHubsConnectionName": "EventHubsConnection"
            },
            "LogLevelLimit": "Information",
            "StorageLogLevelLimit": "Warning",
            "TransportLogLevelLimit": "Warning",
            "EventLogLevelLimit": "Warning",
            "WorkItemLogLevelLimit": "Warning",
            "ClientLogLevelLimit": "Warning",
            "LoadMonitorLogLevelLimit": "Warning",
            "TraceToConsole": false,
            "TraceToBlob": false,
            "InstanceCacheSizeMB": "2048",
            "CacheOrchestrationCursors": true
        }
    }
}
```

I tied bumping `InstanceCacheSizeMB` up to 2GB, got a bit further, but it died again. If the answer is to increase memory and/or limit concurrency, that's fine, but the fact that it puts the host in this weirdly unrecoverable state is my biggest concern.

Open to suggestions and happy to provide additional detail if needed. 

Thank you!

**Comment by sebastianburckhardt** [comment]
There are several things going on here. To avoid out of memory, usually lowering (not raising) the InstanceCacheSize is the thing to do (because less stuff is kept in memory that way). 

But as you say, the really troublesome one is the failure to recover the partition which should never happen, and is something we need to fix.

```
Part04 !!! failed to start partition in EventHubsProcessor.StartPartitionAsync: System.ArgumentException: Destination array was not long enough. Check the destination index, length, and the array's lower bounds. (Parameter 'destinationArray')
at System.Array.Copy(Array sourceArray, Int32 sourceIndex, Array destinationArray, Int32 destinationIndex, Int32 length, Boolean reliable)
at System.Array.Copy(Array sourceArray, Array destinationArray, Int32 length)
at FASTER.core.FasterKV`2.SetRecoveryPageRanges(HybridLogCheckpointInfo recoveredHLCInfo, Int32 numPagesToPreload, Int64 fromAddress, Int64& tailAddress, Int64& headAddress, Int64& scanFromAddress)
at FASTER.core.FasterKV`2.InternalRecoverAsync(IndexCheckpointInfo recoveredICInfo, HybridLogCheckpointInfo recoveredHLCInfo, Int32 numPagesToPreload, Boolean undoNextVersion, Int64 recoverTo, CancellationToken cancellationToken)
```

this seems to indicate a bug somewhere low in the stack. @badrishc, any chance you recognize this error?

**Comment by sebastianburckhardt** [comment]
I have seen this in my own deployment while testing the fix for #231. So this is confirmed as a bug. Needs more investigation.

**Comment by nls-jbarntish** [comment]
Thanks @sebastianburckhardt for the super quick look at this.

> To avoid out of memory, usually lowering (not raising) the InstanceCacheSize is the thing to do (because less stuff is kept in memory that way).

What you're saying makes sense, but I did have it set to `500` at one point and it bailed almost immediately. Didn't execute a single sub orchestration. That said, if  memory serves, the func app didn't fall over in the process and remained responsive. But I could be wrong about that; we've been poking at this for a while trying to tease out the correct setup.

Worth mentioning: I just tried a run with with `"CacheOrchestrationCursors": false`. This completed successfully (yay!), with only 2 exception logs (which are likely bugs on our end). With that setting off it still gives us the performance we're hoping for with some room to grow because the func app did not appear to be maxed.

What's interesting is this run with caching off reduced the # of SQL workers significantly. Maybe that's a just a result of less throughput. Or, another theory is if caching keeps activity instances hot, its related to connection pooling? Each instance would have it's own pool. Not sure how much memory that chews, but I think that would certainly add up. 

![image](https://user-images.githubusercontent.com/105935894/225431188-1a61d22d-0a6f-4509-946e-f9100b2af27b.png)

**Comment by sebastianburckhardt** [comment]
The bug that caused recovery to fail is now being tracked in a separate issue #234. 

I will keep this issue open until we can confirm that there aren't any anomalies with the memory consumption that you observed.

**Comment by sebastianburckhardt** [comment]
A fix for the bug was released with 1.3.3.

Let me know if you are still experiencing issues.

BTW, disabling the `CacheOrchestrationCursors` setting seems like a very good idea in situations where you want to reduce memory useage.

**Comment by jbarntish** [comment]
We've upgraded to the latest a few weeks ago and haven't encountered any more memory issues. Will land another issue if we see anything similar.
